### DISCUSSION TITLE
Image Resizer doesn't work when Winter CMS is installed in a subdirectory

### DISCUSSION THREAD

**Discussion by bugzbrown** [wintercms/winter#213] [Q&A] [answered]
- Winter CMS Build: 1.1.3
- PHP Version:
- Database Engine: mysql
- Plugins Installed: 

### Description:
When Winter is installed as a subdirectory and you try to use resizer to resize images in your theme or component, you get a 404 because it tries to find the URL /resizer/d2...cf1/%252Fstorage%252Fapp%252....

### Steps To Reproduce:
1) install wintercms as a subdirectory of a site.
2) try to resize an image on the theme using the `<img src="{{ 'image.jpg'  | media | resize(300,300) }}" />`

### Suggested Fix:
I identified that `modules/system/classes/imageResizer.php` has two points where it returns an URL without adding the app path:

in **public function getResizerUrl** (line ~502)
and sequentially where it returns an already resized image in **public function getResizedUrl()** (line ~532)

```php
/**
     * Get the URL to the system resizer route for this instance's configuration
     *
     * @return string $url
     */
    public function getResizerUrl()
    {
        // Slashes in URL params have to be double encoded to survive Laravel's router
        // @see https://github.com/octobercms/october/issues/3592#issuecomment-671017380
        $resizedUrl = rawurlencode(rawurlencode($this->getResizedUrl()));

        // Get the current configuration's identifier
        $identifier = $this->getIdentifier();

        // Store the current configuration
        $this->storeConfig();

        // IF YOU ADD THE CONFIG::GET APP.URL IT WORKS
        $url = Config::get('app.url','')."/resizer/$identifier/$resizedUrl";

        if (Config::get('cms.linkPolicy', 'detect') === 'force') {
            $url = Url::to($url);
        }

        return $url;
    }

    /**
     * Get the URL to the resized image
     *
     * @return string
     */
    public function getResizedUrl()
    {
        $url = '';

        if ($this->image['source'] === 'filemodel') {
            $model = $this->getFileModel();
            $thumbFile = $model->getThumbFilename($this->width, $this->height, $this->options);
            $url = $model->getPath($thumbFile);
        } else {
            $resizedDisk = Storage::disk(Config::get('cms.storage.resized.disk', 'local'));
            $url = $resizedDisk->url($this->getPathToResizedImage());
        }

        // Ensure that a properly encoded URL is returned
        $segments = explode('/', $url);
        $lastSegment = array_pop($segments);

        // AGAIN HERE - ADDING THE APPURL WORKS TOO
        $url = Config::get('app.url','').implode('/', $segments) . '/' . rawurlencode(rawurldecode($lastSegment));

        if (Config::get('cms.linkPolicy', 'detect') === 'force') {
            $url = Url::to($url);
        }

        return $url;
    }
```

**Comment by bennothommo** [comment]
Thanks @bugzbrown. Would you be willing to put this in a PR for us?

**Comment by arvislacis** [comment]
@bugzbrown @bennothommo Actually this may not be the issue because situations like this are documented in https://github.com/wintercms/winter/blob/4f953107ce462e07d24d5fea35aa1d08d7e047aa/config/cms.php#L303

**Comment by LukeTowers** [comment]
@bugzbrown you should be able to either add the subfolder to the path or set a linkPolicy of force and it should work. Give that a try and report back please.

**Comment by arvislacis** [comment]
In my opinion this may be closed or at least "bug" label should be replaced with something else if there are plans to improve this in future. Currently this can be fixed by adjusting config file.

**Comment by LukeTowers** [comment, marked as answer]
> If you have installed Winter in a subfolder, are using local storage and are not using a linkPolicy of 'force' you should include the path to the subfolder in the `path` option for these storage configurations.

See https://github.com/wintercms/winter/blob/4f953107ce462e07d24d5fea35aa1d08d7e047aa/config/cms.php#L303-L306

**Comment by der-On** [comment]
I have a slightly similar issue. I have installed winter in a subdir `wintercms` and set my app url to `http://localhost/wintercms` (locally), and added it to the paths however the `/resizer/...` urls are all generated without the base url in front. If I use link policy `force` it will generate wrong resized image urls: `http://wintercms/wintercms/...` as the base url and the resized image urls also contain the subfolder twice, even if removing it from the path of the resized storage. My workaround for now is to always convert `/resizer/...` urls to absolute urls in ImageResizer::getResizerUrl() either using Url::to when link policy is force or using `url()` helper when link policy is not force.

**Comment by croaker000** [comment]
While this is still a problem (even for the "Media" tool in the CMS), I'm using this solution:
1. Add an `APP_PATH='mysubdir'` entry to .env and point it to the Winder install sub-directory
2. Per the "official" answer [https://github.com/orgs/wintercms/discussions/213#discussioncomment-899632], edit the cms.php lines:
`'path' => env('APP_FOLDER','').'/storage/app/uploads'` etc...
3. Update ./modules/system/classes/ImageResizer.php changing line 593 to:
`$url = env('APP_FOLDER','')."/resizer/$identifier/$resizedUrl";`
